### PR TITLE
Bird sighting

### PR DESCRIPTION
**Pull request by kimbambala**
Most basic functionalities are done and ready to be merged with database-rework.

### CODE REVIEW

File: dotnet/Capstone/Capstone.csproj [hunk 1]
```diff
@@ -6,6 +6,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="MaxMind.GeoIP2" Version="5.1.0" />
```
Comment (ahdisease) [inline dotnet/Capstone/Capstone.csproj:9]: Do you know where this came from? If so, what does it do?

File: dotnet/Capstone/Controllers/BirdController.cs [hunk 2]
```diff
@@ -71,35 +87,117 @@ public Bird getRandomBird()
...
+                result = BadRequest(new { message = errorMessage });
             }
 
+            return result;
+
         }
 
-        [HttpPut("/birds/{id}")]
-        public void editBird(Bird updatedBird, int id)
+        [HttpPut("/updateBird/{id}")]
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdController.cs]: This id in this route needs removed. See format in [BirdService.js](https://github.com/ahdisease/Bird-Finder/blob/bird-sighting/vue/src/services/BirdService.js)

File: dotnet/Capstone/Controllers/BirdController.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using Microsoft.AspNetCore.Mvc;
 using System;
 using System.Collections.Generic;
+using System.Net.Mail;
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdController.cs]: Is this being used?

File: dotnet/Capstone/Controllers/BirdController.cs [hunk 4]
```diff
@@ -13,10 +14,12 @@ namespace Capstone.Controllers
     public class BirdController : ControllerBase
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdController.cs:14]: Methods in this file are generally printing to the console rather than returning an error. I'm good to correct this in the database-rework branch, but I wanted to acknowledge that.

File: dotnet/Capstone/Controllers/BirdListController.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            if (birdList == null)
+            {
+                Console.WriteLine("No lists");
+            }
+
+            return birdList;
+        }
+
+
+        [HttpPut("/editList/{listId}")]
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdListController.cs]: Remove id from path.

File: dotnet/Capstone/Controllers/BirdListController.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        public IActionResult editList(BirdList list, int listId)
+        {
+            const string errorMessage = "An error occurred and list could not be modified.";
+            
+            IActionResult result;
+
+            try
+            {
+                birdListDao.editList(list, listId);
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdListController.cs:79]: Nothing prevents a user from changing another user's list. Only the name is effected by this, so I'm good to save it for the database-rework to main merge

File: dotnet/Capstone/Controllers/BirdController.cs (hunk 4 above)
Comment (ahdisease) [reply dotnet/Capstone/Controllers/BirdController.cs:14]: I created an [issue](https://github.com/ahdisease/Bird-Finder/issues/8) for us to remember to correct this

File: dotnet/Capstone/Controllers/BirdListController.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            catch
+            {
+                result = BadRequest(new { message = errorMessage });
+            }
+            return result;
+
+        }
+
+
+        [HttpDelete("/deleteList/{listId}")]
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdListController.cs:95]: User should be verified to confirm list belongs to them.

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+using System;
+using Capstone.Models;
+using Capstone.Exceptions;
+using System.Security.Principal;
+
+namespace Capstone.Controllers
+{
+    [Route("[controller]")]
+    [ApiController]
+    public class BirdNoteController : ControllerBase
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs:14]: The following methods in BirdNoteController print console strings instead of returning error messages:
- getNotes
- getNote

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        private readonly IBirdNoteDao birdSightingDao;
+        public BirdNoteController(IBirdNoteDao birdSightingDao)
+        {
+            this.birdSightingDao = birdSightingDao;
+
+        }
+
+        [HttpGet("/bird/{birdId}/notes")]
+        public List<BirdNote> getNotes(int birdId)
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs]: Should this limit access to notes owned by user?

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            List<BirdNote> sightingList = birdSightingDao.getSightings(birdId);
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs]: Should this limit access to notes owned by user?

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                Console.WriteLine("No bird");
+            }
+
+            return sightingList;
+
+        }
+
+        [HttpGet("/note")]
+        public BirdNote getNote(int id)
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs]: Should this limit access to notes owned by user?

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            if (sighting == null)
+            {
+                Console.WriteLine("No bird");
+            }
+
+            return sighting;
+
+        }
+
+        [HttpPost("/newNote/{birdId}")]
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs]: I thought we removed this id in the path? It needs removed to match BirdNoteService.js

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        public IActionResult addNote([FromBody] BirdNote newBirdSighting)
+        {
+            const string errorMessage = "An error occurred and a bird was not created.";
+
+            IActionResult result;
+
+            DateTime today = DateTime.Now;
+
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs]: Where is our date validation?

Comment (ahdisease) [reply dotnet/Capstone/Controllers/BirdNoteController.cs]: Nevermind, I see we only adjusted the editNote method. Enter this code here:
```suggestion
    int currentDate = sighting.DateSpotted.CompareTo(DateTime.Now);

        if (currentDate >= 0)
        {
            return BadRequest(new { message = "Cannot enter future date" });
        }     
```

File: dotnet/Capstone/Controllers/BirdNoteController.cs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                result = BadRequest(new { message = errorMessage });
+            }
+
+
+            return result;
+        }
+         
+        [HttpDelete("/deleteNote/{id}")]
+        public IActionResult deleteNote(int id)
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdNoteController.cs:169]: Nothing prevents a user from deleting another user's note

File: dotnet/database/capstone.sql [hunk 15]
```diff
@@ -99,17 +99,17 @@ UPDATE users
 		--id 1
 INSERT INTO bird (list_id, name, picture, zip_code) VALUES ( 1, 'Blue Jay', 'https://macaulaylibrary.org/asset/609892314', '11111'); 
 		--id 2
-INSERT INTO bird (list_id, name, picture, zip_code) VALUES ( 1, 'Canary', 'https://images.unsplash.com/photo-1586788454110-2d7a4382bd01?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1974&q=80', '11111');
+INSERT INTO bird (list_id, name, picture, zip_code) VALUES ( 1, 'Canary', 'https://macaulaylibrary.org/asset/609854044', '11111');
```
Comment (ahdisease) [inline dotnet/database/capstone.sql:102]: Thanks for these

File: dotnet/Capstone/DAO/UserProfileSqlDao.cs [hunk 16]
```diff
@@ -90,7 +90,7 @@ public UserProfile UpdateUserProfile(UserProfile profile, string username)
             {
                 sqlSet += ", users.skill_level = @skill_level";
             }
-            if (profile.FavoriteBird.id > 0)
+            if (profile.FavoriteBird.BirdId > 0)
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/UserProfileSqlDao.cs]: This is a good place to remind you to merge the updated database-rework into this branch. This specific line change may cause a small merge conflict, FYI.

File: dotnet/Capstone/Controllers/BirdNoteController.cs (hunk 11 above)
Comment (ahdisease) [reply dotnet/Capstone/Controllers/BirdNoteController.cs]: Actually, looking at the BirdNoteService.js, there's no id being passed. That said, it's unclear what the intention of this method was, and there are no uses of this route in the supplied front end.

File: dotnet/Capstone/DAO/BirdSqlDao.cs [hunk 17]
```diff
@@ -4,6 +4,7 @@
 using System.Collections.Generic;
 using System.Data;
 using System.Data.SqlClient;
+using System.Reflection.Emit;
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdSqlDao.cs]: Why was this added?

File: dotnet/Capstone/DAO/BirdSqlDao.cs [hunk 18]
```diff
@@ -12,21 +13,54 @@ namespace Capstone.DAO
...
+        public Bird createBird(Bird bird, int listId)
         {
             Bird newBird = null;
 
-            String sql = "INSERT INTO bird(name, description, picture) " +
-                "VALUES(@name, @description, @picture)";
+            String sql = "INSERT INTO bird(list_id, name, picture, zip_code) " +
+                "VALUES(@list_id, @name, @picture, @zip_code)";
 
             int newBirdId = 0;
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdSqlDao.cs:64]: newBirdId is assigned 0 but isn't used. Did you intend to return this number from your SQL statement? If not, it doesn't seem to be serving any purpose.

Edit: Looks like your ExecuteScalar method might be applying to this? I'd love a talk through sometime.

File: dotnet/Capstone/DAO/BirdSqlDao.cs [hunk 19]
```diff
@@ -199,11 +265,12 @@ public List<Bird> listAllBirds()
...
-            bird.id = Convert.ToInt32(reader["id"]);
-            bird.name = Convert.ToString(reader["name"]);
-            bird.description = Convert.ToString(reader["description"]);
-            bird.imgUrl = Convert.ToString(reader["picture"]);
-         
+            bird.BirdId = Convert.ToInt32(reader["id"]); 
+            bird.ListId = Convert.ToInt32(reader["list_id"]);
+            bird.BirdName = Convert.ToString(reader["name"]);
+            bird.ImgUrl = Convert.ToString(reader["picture"]);
+            bird.ZipCode = Convert.ToString(reader["zip_code"]);
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdSqlDao.cs:271]: Because these items are not marked "NOT NULL" in the schema, there's a chance for them to cause errors when pulling a null value (I believe SQL Server null values aren't innately converted to C# null values). See the [GetSafeString](https://github.com/ahdisease/Bird-Finder/blob/654e710664ed8688628956d2bd7b0a53f0878d21/dotnet/Capstone/DAO/UserProfileSqlDao.cs#L238) method in UserProfileSqlDao for an example of how to handle this.

File: dotnet/Capstone/DAO/BirdDao.cs
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdDao.cs:1]: It's standard in C# to name interfaces with an I prefix.

[For reference](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces)

File: dotnet/Capstone/DAO/BirdListSqlDao.cs [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+                newBirdList = getList(newListId);
+            }
+            catch(SqlException ex)
+            {
+                throw new DaoException("SQL exception error", ex);
+            }
+            return newBirdList;
+        }
+
+        public void deleteList(int listId)
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdListSqlDao.cs:52]: This method could probably use a special handling case to check if there are any birds in the list before deleting, or an Exception that is thrown in that outcome.

File: dotnet/Capstone/DAO/BirdNoteSqlDao.cs [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+        {
+            BirdNote birdSighting = new BirdNote();
+            birdSighting.NoteId = Convert.ToInt32(reader["id"]);
+            birdSighting.BirdId = Convert.ToInt32(reader["bird_id"]);
+            birdSighting.DateSpotted = Convert.ToDateTime(reader[("date_sighted")]);
+            birdSighting.NumMales = Convert.ToInt32(reader["males_spotted"]);
+            birdSighting.NumFemales = Convert.ToInt32(reader["females_spotted"]);
+            birdSighting.FeederType = Convert.ToString(reader["feeder_type"]);
+            birdSighting.FoodBlend = Convert.ToString(reader["food_blend"]);
+            birdSighting.Notes = Convert.ToString(reader["notes"]);
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdNoteSqlDao.cs:181]: Because these items are not marked "NOT NULL" in the schema, there's a chance for them to cause errors when pulling a null value. See the [GetSafeString](https://github.com/ahdisease/Bird-Finder/blob/654e710664ed8688628956d2bd7b0a53f0878d21/dotnet/Capstone/DAO/UserProfileSqlDao.cs#L238) method in UserProfileSqlDao for an example of how to handle this.

File: dotnet/Capstone/DAO/BirdNoteSqlDao.cs [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+        }
+
+        private BirdNote MapRowToBirdSighting(SqlDataReader reader)
+        {
+            BirdNote birdSighting = new BirdNote();
+            birdSighting.NoteId = Convert.ToInt32(reader["id"]);
+            birdSighting.BirdId = Convert.ToInt32(reader["bird_id"]);
+            birdSighting.DateSpotted = Convert.ToDateTime(reader[("date_sighted")]);
+            birdSighting.NumMales = Convert.ToInt32(reader["males_spotted"]);
+            birdSighting.NumFemales = Convert.ToInt32(reader["females_spotted"]);
```
Comment (ahdisease) [inline dotnet/Capstone/DAO/BirdNoteSqlDao.cs:178]: Integers also have the potential to be null in SQL Server. See the [GetSafeInt](https://github.com/ahdisease/Bird-Finder/blob/654e710664ed8688628956d2bd7b0a53f0878d21/dotnet/Capstone/DAO/UserProfileSqlDao.cs#L247) method in UserProfileSqlDao for an example of how to handle this.

**Review by ahdisease — CHANGES_REQUESTED** [review]
I've checked over the new and updated files and in most cases the comments I've made are quick fixes. 

Because we're merging this into another working branch, most changes can be left unedited to fix in the PR to merge database-rework, but I'd ask you create an Issue from my comment when you do. That should prevent us forgetting to address these errors and upgrades, and keeps us from having to do quite as in depth of a review, since all of the code has been looked over already.

The exception to pushing the work until later would be that you _should_ pull the latest version of database-rework and merge the new changes from PR #6 into bird-sighting. After you've done that, I can change my approval status.

File: dotnet/Capstone/Capstone.csproj (hunk 1 above)
Comment (kimbambala) [reply dotnet/Capstone/Capstone.csproj:9]: I have no idea.

File: dotnet/Capstone/Controllers/BirdController.cs (hunk 2 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdController.cs]: Done

File: dotnet/Capstone/Controllers/BirdController.cs (hunk 3 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdController.cs]: No. I took it off.

File: dotnet/Capstone/Controllers/BirdListController.cs (hunk 5 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdListController.cs]: Done

File: dotnet/Capstone/DAO/BirdSqlDao.cs (hunk 17 above)
Comment (kimbambala) [reply dotnet/Capstone/DAO/BirdSqlDao.cs]: No sure. I just took it off.

File: dotnet/Capstone/DAO/BirdSqlDao.cs (hunk 18 above)
Comment (kimbambala) [reply dotnet/Capstone/DAO/BirdSqlDao.cs:64]: This is just declaring the new bird id which will be later set to the newly created bird object.

File: dotnet/Capstone/DAO/BirdDao.cs
Comment (kimbambala) [reply dotnet/Capstone/DAO/BirdDao.cs:1]: Done. I learned this after and forgot to change this one. This was the first Interface I created.

File: dotnet/Capstone/DAO/BirdListSqlDao.cs (hunk 20 above)
Comment (kimbambala) [reply dotnet/Capstone/DAO/BirdListSqlDao.cs:52]: What will be the use of such handling given that the method is deleting the list whether it has birds or not.

File: dotnet/Capstone/Controllers/BirdNoteController.cs (hunk 8 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdNoteController.cs:14]: Done

File: dotnet/Capstone/Controllers/BirdNoteController.cs (hunk 11 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdNoteController.cs]: I can't see a use for it as well especially without an id being passed. I will still leave it and maybe we will find a its use later.

File: dotnet/Capstone/Controllers/BirdNoteController.cs (hunk 12 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdNoteController.cs]: Done

File: dotnet/Capstone/Controllers/BirdNoteController.cs (hunk 13 above)
Comment (kimbambala) [reply dotnet/Capstone/Controllers/BirdNoteController.cs]: Done

**Comment by kimbambala** [comment]
The latest version of database-rework has been pulled and merged with bird-finder.

File: dotnet/Capstone/Controllers/BirdController.cs [hunk 23]
```diff
@@ -48,7 +48,8 @@ public IActionResult ListAllBirds()
                 }
                 else
                 {
-                    result = NotFound(new { message = errorMessage });
+                    result = result = NotFound(new { message = errorMessage });
```
Comment (ahdisease) [inline dotnet/Capstone/Controllers/BirdController.cs:51]: Is this a fix? Did you mean to return the NotFound?

**Comment by ahdisease** [comment]
After addressing [this](https://github.com/ahdisease/Bird-Finder/pull/7#discussion_r1371102981) comment, we're good to merge with the database-rework branch. The issue isn't actually preventing the code from compiling, but just looks strange.

I'll convert remaining comments to issues, and add another: BirdController, BirdListController, and BirdNoteController don't require a user to be authorized.

**Review by ahdisease — APPROVED** [review]
